### ISSUE TITLE
pyright is not able to infer the correct type when there are two possibilities

### ISSUE THREAD

**Issue by renzhentaxi** [microsoft/pyright#913]
**Describe the bug**
This is similar to this  #910 . I am expecting pyright to resolve a typevar to `str` but it is not. Im not sure if this is a bug or if it is working as intended. If this is working as intended, is there a way to work around this?

**To Reproduce**
```python
from typing import Union, List, TypeVar, Type

T = TypeVar('T')


def to_list(obj_type: Type[T], obj: Union[List[T], T]) -> List[T]:
    return []
```
**Expected behavior**

If I pass str, it works
```python
input_str: str = "string"
output_list: List[str] = to_list(str, input_str)
```
If I pass a list of str, it doesn't work
```python
input_list: List[str] = ["string"]
output_list:List[str] = to_list(str, input_list)
```
I receive an error message:

>  11:17 - error: Expression of type "List[str | List[str]]" cannot be assigned to declared type "List[str]"
  TypeVar "_T" is invariant
    Type "str | List[str]" cannot be assigned to type "str" (reportGeneralTypeIssues)

It seems that it resolved `T` to `Union[List[str], str]`. However, shouldn't it resolve `T` to `str` since I passed `str` to `obj_type`?

**VS Code extension or command-line**
This issue occurs when I run pyright directly or when I using the pylance extension.
pyright version: pyright 1.1.59
pylance version: v2020.7.4

**Comment by erictraut** [comment]
Thanks for submitting the issue.

It's not clear to me that this is actually a "bug", but I can understand why you would see it as such. When a type checker "solves" for a set of type variables, there may be more than one answer that satisfies all the constraints. This is one such case, and Pyright is finding one solution, but you're expecting another. It should arguably find the "simplest" solution. That's easier said than done. It's not always clear which solution is "simpler", especially when multiple type variables are involved. I'll need to think about this more.

**Comment by erictraut** [comment]
I implemented a "complexity scoring" mechanism in the type var solver. When it encounters a union, it now attempts to solve the type variables for each of the elements in the union and picks the "most complete but least complex" of the results. In other words, it favors solutions that provide type var mappings for the greatest number of type variables but also favors simpler solutions. This produces the same answer regardless of the order in which the union elements are defined. For example, you used `Union[List[T], T]]` in your code above, but it should produce the same answer if you use `Union[T, List[T]]`. The "complexity score" is based on some heuristics that may need tuning.

This will be in the next version of Pyright and Pylance.

**Comment by erictraut** [comment]
This is addressed in 1.1.60, which I just published. It will also be in the next version of Pylance.
